src/main/java/org/apache/mesos/hdfs/Scheduler.java

```diff
@@ -191,197 +363,138 @@ private ExecutorInfo createExecutor(String taskIdName, String nodeName, String e
+          HDFSConstants.NAME_NODE_EXECUTOR_ID);
+      return true;
     }
+    return false;
   }
 
-  @Override
-  synchronized public void resourceOffers(SchedulerDriver driver, List<Offer> offers) {
-    log.info(String.format("Received %d offers", offers.size()));
-    //TODO(elingg) all datanodes can be launched together after the other nodes have initialized.
-    //Remove this waiting period for datanodes.
-    if (!liveState.getStagingTasks().isEmpty()) {
-      log.info("Declining offers because tasks are currently staging");
-      for (Offer offer : offers) {
-        driver.declineOffer(offer.getId());
-      }
-      return;
+  private boolean maybeLaunchDataNode(SchedulerDriver driver, Offer offer) {
+    if (persistentState.dataNodeRunningOnSlave(offer.getHostname()) ||
+        persistentState.nameNodeRunningOnSlave(offer.getHostname()) ||
+        persistentState.journalNodeRunningOnSlave(offer.getHostname())) {
+      log.info(String.format("Already running hdfs task on %s", offer.getHostname()));
+    } else if (offerNotEnoughResources(offer, conf.getDataNodeCpus(), conf.getDataNodeHeapSize())) {
+      log.info("Offer does not have enough resources");
+    } else {
+      launchNode(
+          driver,
+          offer,
+          HDFSConstants.DATA_NODE_ID,
+          Arrays.asList(HDFSConstants.DATA_NODE_ID),
+          HDFSConstants.NODE_EXECUTOR_ID);
+      return true;
     }
+    return false;
+  }
 
-    if (liveState.getNameNodes().size() == 0 && liveState.getJournalNodes().size() == 0) {
-      log.info("No NameNodes or JournalNodes found.  Collecting offers until we have sufficient "
-          + "capacity to launch.");
-      for (Offer offer: offers) {
-          pendingOffers.put(offer.getId(), offer);
-      }
-        
-      if (!initializingCluster) {
-        log.info(String.format("Launching initial nodes with %d pending offers",
-            pendingOffers.size()));
-        initializingCluster = true;
-        launchInitialJournalNodes(driver, pendingOffers.values());
-        launchInitialNameNodes(driver, pendingOffers.values());
-        // Decline any remaining offer
-        for (OfferID offerID : pendingOffers.keySet()) {
-          driver.declineOffer(offerID);
-        }
-        pendingOffers.clear();
-      }
-      return;
-    }
+  private void sendMessageTo(SchedulerDriver driver, TaskID taskId, SlaveID slaveID, String message) {
+    log.info(String.format("Sending message '%s' to taskId=%s, slaveId=%s", message,
+        taskId.getValue(), slaveID.getValue()));
+    String postfix = taskId.getValue();
+    postfix = postfix.substring(postfix.indexOf(".") + 1, postfix.length());
+    postfix = postfix.substring(postfix.indexOf(".") + 1, postfix.length());
+    driver.sendFrameworkMessage(
+        ExecutorID.newBuilder().setValue("executor." + postfix).build(),
+        slaveID,
+        message.getBytes());
+  }
 
```

Comment:
Could we add these methods related to state to a separate class? I'd like to avoid growing the scheduler too big. Breaking down the logic in an object oriented manner would be great.